File: apps/vibe-trading/launch.py
#!/usr/bin/env python3
"""UnifiedApp marketplace service entrypoint for Vibe-Trading (`node-service` kind).

The UnifiedApp desktop host (Tauri `desktop_core`) spawns this exactly the way it
spawns OpenDesign's Node daemon — `service.command` is a generic executable, so a
Python interpreter is fine:

    manifest.json:
      "service": {
        "command": "runtime/python/bin/python3",
        "args": ["launch.py"],
        "healthPath": "/health",
        ...
      }

This launcher is the Python analogue of OpenDesign's hand-vendored broker client
(`apps/daemon/src/unified-auth.ts`): it exchanges the loopback broker secret for a
short-lived, app-scoped UnifiedAI token and points Vibe-Trading's OpenAI-compatible
LLM client (langchain-openai) at the Unified gateway. The user's long-lived
credential never reaches this process.

Host-injected env (see desktop_core/src/service.rs + broker.rs):
  OD_PORT                   free port the host allocated — we MUST bind it
  OD_BIND_HOST              127.0.0.1
  OD_DATA_DIR               writable per-app data dir (install root is read-only)
  NO_OPEN                   "1" (the desktop opens the URL itself)
  UNIFIED_APP_SLUG          our marketplace slug ("vibe-trading")
  UNIFIED_BROKER_URL        loopback broker base url        (broker path)
  UNIFIED_BROKER_TOKEN      per-launch shared secret        (x-broker-token header)
  UNIFIED_HOST_TRUST_TOKEN  proof we were started by the trusted host

`UNIFIED_API_URL` is deliberately NOT injected by the host — we default it to the
local dev gateway (loopback:3141, where the desktop runs unified-api) and allow an
override (e.g. https://api.unifiedai.app for a non-dev gateway). This mirrors the
Hermes bundle's launcher.
"""

from __future__ import annotations

import json
import os
import sys
import threading
import time
import urllib.error
import urllib.request
from pathlib import Path

BUNDLE_ROOT = Path(__file__).resolve().parent
AGENT_DIR = BUNDLE_ROOT / "agent"
FRONTEND_DIST = BUNDLE_ROOT / "frontend" / "dist"

# Unified gateway, OpenAI-compatible surface mounted under /api/v1 (chat/completions,
# models, embeddings, images, audio). The desktop runs unified-api on loopback:3141 and
# does NOT inject a gateway URL (see the UNIFIED_API_URL note above), so apps default it
# here. Override with UNIFIED_API_URL for a non-dev gateway (e.g. https://api.unifiedai.app).
DEFAULT_UNIFIED_API_HOST = "http://localhost:3141"
# `auto` lets the gateway pick a concrete model; the user can change it in-app
# (Settings → LLM) afterwards. LANGCHAIN_MODEL_NAME is required or build_llm raises.
DEFAULT_MODEL = "auto"


def _log(msg: str) -> None:
    print(f"[vibe-launch] {msg}", flush=True)


def _unified_api_base() -> str:
    """OpenAI-compatible base URL langchain-openai should target."""
    host = (os.environ.get("UNIFIED_API_URL") or DEFAULT_UNIFIED_API_HOST).rstrip("/")
    # Accept either a bare host or one that already includes the /api/v1 suffix.
    return host if host.endswith("/api/v1") else f"{host}/api/v1"


def _broker_coords() -> tuple[str, str]:
    """Current loopback-broker (url, shared-secret).

    Read from OD_DATA_DIR/.broker.json (which the desktop host rewrites on every
    spawn AND reuse) first, falling back to the env injected at spawn time. This is
    what lets a daemon REUSED across a desktop restart — or kept alive by a sibling
    process that has since quit (cross-process shell⇄solo reuse) — keep re-minting
    against the LIVE broker. The broker URL+secret rotate per desktop launch, so a
    reused daemon's *env* coords go stale and minting fails forever ("Connection
    refused" → the app token expires → every model call 401s "Invalid or expired
    JWT"). The host rewrites the file with the live coords, so the next refresh
    self-heals (~30s) without restarting the daemon.
    """
    data_dir = os.environ.get("OD_DATA_DIR", "")
    if data_dir:
        try:
            raw = json.loads((Path(data_dir) / ".broker.json").read_text())
            url = (raw.get("url") or "").rstrip("/")
            tok = raw.get("token") or ""
            if url and tok:
                return url, tok
        except (OSError, ValueError):
            pass  # missing/partial file → fall back to the spawn-time env
    return (
        os.environ.get("UNIFIED_BROKER_URL", "").rstrip("/"),
        os.environ.get("UNIFIED_BROKER_TOKEN", ""),
    )


def _mint_token() -> tuple[str | None, int]:
    """Exchange the broker shared-secret for a short-lived app-scoped token.

    Mirrors desktop_core/src/broker.rs: POST {broker_url}/token with header
    `x-broker-token` and JSON body {"app_slug": slug}. Returns (token, expires_in),
    or (None, 0) when no broker is present (running outside the desktop). Broker
    coords are re-read every call (see `_broker_coords`) so a rotated broker is
    picked up without a restart.
    """
    broker_url, broker_tok = _broker_coords()
    slug = os.environ.get("UNIFIED_APP_SLUG", "vibe-trading")
    if not broker_url or not broker_tok:
        return None, 0
    req = urllib.request.Request(
        f"{broker_url}/token",
        data=json.dumps({"app_slug": slug}).encode(),
        method="POST",
        headers={"content-type": "application/json", "x-broker-token": broker_tok},
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
        token = data.get("token")
        return (token, int(data.get("expires_in") or 300)) if token else (None, 0)
    except (urllib.error.URLError, ValueError, OSError) as exc:
        _log(f"broker token mint failed: {exc}")
        return None, 0


def _apply_token(token: str) -> None:
    """Point the UnifiedAI provider — and the generic OPENAI_* vars langchain-openai
    actually reads — at the gateway with the freshly minted token."""
    base = _unified_api_base()
    os.environ["UNIFIED_API_BASE_URL"] = base
    os.environ["UNIFIED_API_KEY"] = token
    # llm.py build_llm() re-reads these on every call, so updating them in-process
    # refreshes the token for subsequent agent runs without a restart.
    os.environ["OPENAI_API_KEY"] = token
    os.environ["OPENAI_BASE_URL"] = base
    os.environ["OPENAI_API_BASE"] = base


def _write_host_token(token: str) -> None:
    """The host reads OD_DATA_DIR/.token back after health-check to hand to the
    webview (parity with OpenDesign's `read_app_token`). Best-effort."""
    data_dir = os.environ.get("OD_DATA_DIR", "")
    if not data_dir:
        return
    try:
        d = Path(data_dir)
        d.mkdir(parents=True, exist_ok=True)
        (d / ".token").write_text(token)
    except OSError as exc:
        _log(f"could not write host token file: {exc}")


def _refresh_loop(expires_in: int) -> None:
    """Re-mint the short-lived (~5 min) app token before it expires. Runs in a
    daemon thread sharing this process's os.environ with the FastAPI app."""
    while True:
        time.sleep(max(30, expires_in - 60))
        token, ttl = _mint_token()
        if token:
            _apply_token(token)
            _write_host_token(token)
            expires_in = ttl
            _log("refreshed Unified app token")
        else:
            # Retry soon (sleep floor is 30s) so a rotated broker — written to
            # .broker.json by the host on restart/reuse — is picked up quickly and
            # the daemon stops 401ing. 90 → next sleep max(30, 90-60) = 30s.
            expires_in = 90


def main() -> int:
    # Make the bundle's app source importable WITHOUT a (non-relocatable) editable
    # install — only third-party deps live in the embedded interpreter.
    sys.path.insert(0, str(AGENT_DIR))

    host = os.environ.get("OD_BIND_HOST", "127.0.0.1")
    port = os.environ.get("OD_PORT", "8899")

    # Relocate writable runtime state + locate the prebuilt frontend, since the
    # install tree is read-only. These MUST be set before importing api_server,
    # which resolves them at module import time.
    os.environ.setdefault("VIBE_TRADING_DATA_DIR", os.environ.get("OD_DATA_DIR", str(AGENT_DIR)))
    if FRONTEND_DIST.exists():
        os.environ.setdefault("VIBE_TRADING_FRONTEND_DIST", str(FRONTEND_DIST))
    # Loopback-only FastAPI auth (empty key = dev/loopback mode). The desktop reaches
    # us over loopback and the SPA is served same-origin, so no API key is needed.
    os.environ.setdefault("API_AUTH_KEY", "")
    os.environ["PYTHONUNBUFFERED"] = "1"

    # We were launched by the UnifiedApp desktop host, so the LLM is "managed":
    # the broker supplies the credential and the in-app settings hide the
    # provider/key controls for an OpenDesign-style zero-config experience
    # (api_server._is_managed reads this flag).
    os.environ.setdefault("VIBE_TRADING_MANAGED", "1")
    # Default to the Unified gateway provider. Kept as a default (not a hard set)
    # so someone running this entrypoint by hand can still override via real env.
    os.environ.setdefault("LANGCHAIN_PROVIDER", "unifiedai")
    os.environ.setdefault("LANGCHAIN_MODEL_NAME", DEFAULT_MODEL)
    os.environ.setdefault("UNIFIED_API_BASE_URL", _unified_api_base())

    token, expires_in = _mint_token()
    if token:
        _apply_token(token)
        _write_host_token(token)
        _log(f"acquired Unified app token (expires_in={expires_in}s) via broker")
        threading.Thread(target=_refresh_loop, args=(expires_in,), daemon=True).start()
    else:
        _log("no broker token (running outside UnifiedApp?); LLM creds left to .env / in-app settings")

    _log(f"serving on {host}:{port} (data={os.environ['VIBE_TRADING_DATA_DIR']})")
    from api_server import serve_main  # imported after env is wired

    return serve_main(["--host", host, "--port", str(port)])


if __name__ == "__main__":
    raise SystemExit(main())

File: apps/activepieces/launch.mjs
#!/usr/bin/env node
/**
 * UnifiedApp marketplace service entrypoint for Activepieces (`node-service` kind).
 *
 * The UnifiedApp desktop host (Tauri `desktop_core`) spawns this as a generic
 * executable per manifest.json:
 *
 *     "service": {
 *       "command": "runtime/node/bin/node",
 *       "args": ["launch.mjs"],
 *       "healthPath": "/api/v1/health",
 *       ...
 *     }
 *
 * Activepieces is a Fastify backend + React/Vite frontend that, in production,
 * serves the built frontend AND the API from a single port (AP_PORT) and runs an
 * out-of-process worker that polls the app over HTTP. We run it in its
 * single-machine configuration — embedded Postgres (PGLITE) + an in-memory queue
 * (MEMORY) — so there are no external Postgres/Redis dependencies.
 *
 * Three jobs:
 *
 *  1. Env bridge. Translate the host's OD_* contract into Activepieces' AP_* env:
 *     bind the host-allocated port, relocate all mutable state under the writable
 *     per-app data dir (the install tree is read-only), and relax CSP
 *     frame-ancestors so the desktop shell can iframe the app.
 *
 *  2. Process model (mirrors Activepieces' own docker-entrypoint.sh single
 *     container). Spawn TWO node children with a shared, stable AP_JWT_SECRET /
 *     AP_ENCRYPTION_KEY:
 *       - app:    packages/server/api/dist/src/main.js   (AP_CONTAINER_TYPE=APP)
 *                 binds AP_PORT, serves UI + API + the job broker.
 *       - worker: packages/server/worker/dist/src/index.js (AP_CONTAINER_TYPE=WORKER_AND_APP
 *                 so it polls http://127.0.0.1:AP_PORT/api/ and starts NO health
 *                 server of its own — avoiding a port clash with the app).
 *     The worker needs AP_WORKER_TOKEN (an HS256 JWT signed with AP_JWT_SECRET),
 *     which we mint exactly as docker-entrypoint.sh does.
 *
 *  3. Gateway auth (best-effort). Exchange the loopback broker secret for a
 *     short-lived, app-scoped UnifiedAI token and run a small loopback proxy that
 *     forwards to the Unified gateway's OpenAI-compatible surface with that token
 *     attached. Point an Activepieces AI provider's "Base URL" at this proxy to
 *     use AI on the user's Unified subscription (no provider API keys, and the
 *     app never sees a long-lived credential). See BUNDLING.md.
 *
 * Host-injected env (see desktop_core/src/service.rs + broker.rs):
 *   OD_PORT                   free port the host allocated — we MUST bind it
 *   OD_BIND_HOST              127.0.0.1
 *   OD_DATA_DIR               writable per-app data dir (install root is read-only)
 *   NO_OPEN                   "1" (the desktop opens the URL itself)
 *   UNIFIED_APP_SLUG          our marketplace slug ("activepieces")
 *   UNIFIED_BROKER_URL        loopback broker base url
 *   UNIFIED_BROKER_TOKEN      per-launch shared secret (x-broker-token header)
 *   UNIFIED_HOST_TRUST_TOKEN  proof we were started by the trusted host
 *
 * UNIFIED_API_URL is deliberately NOT injected by the host — we default it to the
 * production gateway and allow an override (e.g. http://localhost:3141 in dev).
 */

import { spawn } from 'node:child_process'
import crypto from 'node:crypto'
import fs from 'node:fs'
import http from 'node:http'
import https from 'node:https'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const BUNDLE_ROOT = path.dirname(fileURLToPath(import.meta.url))
const log = (msg) => console.log(`[activepieces-launch] ${msg}`)

// ── Host contract ────────────────────────────────────────────────────────────
const BIND_HOST = process.env.OD_BIND_HOST || '127.0.0.1'
const PORT = String(process.env.OD_PORT || '3000')
const DATA_DIR = process.env.OD_DATA_DIR || path.join(BUNDLE_ROOT, '.data')
const SLUG = process.env.UNIFIED_APP_SLUG || 'activepieces'

// Unified gateway, OpenAI-compatible surface mounted at /api/v1. Default matches
// the other bundles (the desktop runs unified-api on loopback:3141 and does NOT
// inject a gateway URL). Override with UNIFIED_API_URL for a non-dev gateway.
const UNIFIED_API_URL = (process.env.UNIFIED_API_URL || 'http://localhost:3141').replace(/\/+$/, '')
const UNIFIED_API_BASE = UNIFIED_API_URL.endsWith('/api/v1') ? UNIFIED_API_URL : `${UNIFIED_API_URL}/api/v1`

// UnifiedApp desktop shell origins permitted to iframe this app (CSP
// frame-ancestors). The macOS production parent origin is the custom scheme
// tauri://localhost, which Activepieces' isValidOrigin() rejects — so this is
// applied via our AP_EMBED_FRAME_ANCESTORS source patch in server.ts, not the
// stock AP_ALLOWED_EMBED_ORIGINS path. Keep in sync with apps/openclaw.
const EMBED_FRAME_ANCESTORS =
  'tauri://localhost http://tauri.localhost https://tauri.localhost http://localhost:1420 http://localhost'

// Stable loopback port for the AI gateway proxy. Fixed (not ephemeral) so a saved
// Activepieces AI-provider base URL survives restarts. Override if it collides.
const AI_PROXY_PORT = Number(process.env.UNIFIED_AI_PROXY_PORT || 25152)

// Writable state dirs under the per-app data dir (install tree is read-only).
const CONFIG_PATH = path.join(DATA_DIR, 'config')
const CACHE_PATH = path.join(DATA_DIR, 'cache')
for (const dir of [DATA_DIR, CONFIG_PATH, CACHE_PATH]) fs.mkdirSync(dir, { recursive: true })

// ── Persisted secrets ────────────────────────────────────────────────────────
// AP_ENCRYPTION_KEY encrypts connection secrets in the PGLite DB; AP_JWT_SECRET
// signs sessions and the worker token. Both MUST be stable across restarts or the
// existing DB can no longer be decrypted / sessions are invalidated.
function loadOrCreateSecrets() {
  const file = path.join(DATA_DIR, 'secrets.json')
  let secrets = {}
  try {
    secrets = JSON.parse(fs.readFileSync(file, 'utf8')) || {}
  } catch {
    /* first run */
  }
  let changed = false
  // 32-char hex (16 bytes) — matches Activepieces' /^[A-Za-z0-9]{32}$/ check.
  if (!secrets.encryptionKey) { secrets.encryptionKey = crypto.randomBytes(16).toString('hex'); changed = true }
  if (!secrets.jwtSecret) { secrets.jwtSecret = crypto.randomBytes(32).toString('hex'); changed = true }
  // Password for the auto-provisioned local admin (server.ts local-autologin). 36
  // hex chars — within Activepieces' 8..64 policy. Persisted so sign-in is stable.
  if (!secrets.autoLoginPassword) { secrets.autoLoginPassword = crypto.randomBytes(18).toString('hex'); changed = true }
  if (changed) {
    try {
      fs.writeFileSync(file, JSON.stringify(secrets), { mode: 0o600 })
    } catch (e) {
      log(`could not persist secrets (will regenerate next launch): ${e}`)
    }
  }
  return secrets
}
const { encryptionKey, jwtSecret, autoLoginPassword } = loadOrCreateSecrets()

// ── Worker token (mirrors docker-entrypoint.sh AP_WORKER_TOKEN) ───────────────
const base64url = (input) =>
  Buffer.from(input).toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
function signWorkerToken() {
  const header = { alg: 'HS256', typ: 'JWT', kid: '1' }
  const iat = Math.floor(Date.now() / 1000)
  const payload = {
    id: crypto.randomUUID(),
    type: 'WORKER',
    iat,
    exp: iat + 100 * 365 * 24 * 60 * 60, // ~100y, like the upstream entrypoint
    iss: 'activepieces',
  }
  const signingInput = `${base64url(JSON.stringify(header))}.${base64url(JSON.stringify(payload))}`
  const signature = base64url(crypto.createHmac('sha256', jwtSecret).update(signingInput).digest())
  return `${signingInput}.${signature}`
}
const workerToken = signWorkerToken()

// ── Shared Activepieces env ───────────────────────────────────────────────────
const sharedEnv = {
  ...process.env,
  NODE_ENV: 'production',
  // Single-machine datastore: embedded Postgres + in-memory queue (no externals).
  AP_DB_TYPE: 'PGLITE',
  AP_REDIS_TYPE: 'MEMORY',
  AP_CONFIG_PATH: CONFIG_PATH,
  AP_CACHE_PATH: CACHE_PATH,
  // 'prod' is REQUIRED for the app to serve the built frontend (server.ts only
  // registers @fastify/static when AP_ENVIRONMENT !== 'dev').
  AP_ENVIRONMENT: 'prod',
  AP_EDITION: 'ce',
  AP_EXECUTION_MODE: 'UNSANDBOXED',
  AP_TELEMETRY_ENABLED: 'false',
  AP_FRONTEND_URL: `http://127.0.0.1:${PORT}`,
  AP_PORT: PORT,
  AP_JWT_SECRET: jwtSecret,
  AP_ENCRYPTION_KEY: encryptionKey,
  AP_WORKER_TOKEN: workerToken,
  // Activepieces' MEMORY queue spawns an EPHEMERAL redis via redis-memory-server,
  // which otherwise downloads the redis binary on first boot (a fragile network
  // dependency that's wiped on every update → ECONNREFUSED until it finishes). Point
  // it at the binary we embed at build time so it never downloads. getSystemPath just
  // access()-checks this, falling back to download if the path is missing.
  REDISMS_SYSTEM_BINARY: path.join(BUNDLE_ROOT, 'runtime', 'redis', 'redis-server'),
  // Consumed by our server.ts patch to allow the desktop shell to iframe the app.
  AP_EMBED_FRAME_ANCESTORS: EMBED_FRAME_ANCESTORS,
  // Local auto-login: provision/sign-in a single local admin and inject its token
  // into index.html so the bundle opens straight to the dashboard — no sign-in
  // screen. AP_LOCAL_AUTOLOGIN gates the authentication-service patch that makes a
  // Community sign-in return a full token (resolving/creating the platform) instead
  // of an onboarding token. A single-user local appliance; clear these to restore login.
  AP_LOCAL_AUTOLOGIN: 'true',
  AP_LOCAL_AUTOLOGIN_EMAIL: 'local@activepieces.local',
  AP_LOCAL_AUTOLOGIN_PASSWORD: autoLoginPassword,
  // Embedded runtimes on PATH: node (this) + bun (the worker shells out to `bun`
  // to install registry pieces at runtime — see worker/.../piece-installer.ts).
  PATH: [
    path.join(BUNDLE_ROOT, 'runtime', 'node', 'bin'),
    path.join(BUNDLE_ROOT, 'runtime', 'bun'),
    process.env.PATH || '',
  ].join(path.delimiter),
}

const NODE_BIN = path.join(BUNDLE_ROOT, 'runtime', 'node', 'bin', 'node')
const APP_ENTRY = path.join('packages', 'server', 'api', 'dist', 'src', 'main.js')
const WORKER_ENTRY = path.join('packages', 'server', 'worker', 'dist', 'src', 'index.js')

// ── Gateway auth + AI proxy ───────────────────────────────────────────────────
let currentToken = null

// Prefer OD_DATA_DIR/.broker.json (the desktop rewrites it on every spawn/reuse)
// over the spawn-time env, so a daemon reused across a desktop restart self-heals
// against rotated broker coords (vibe-trading pattern).
function brokerCoords() {
  try {
    const j = JSON.parse(fs.readFileSync(path.join(DATA_DIR, '.broker.json'), 'utf8'))
    if (j.url && j.token) return { url: String(j.url).replace(/\/+$/, ''), token: j.token }
  } catch {
    /* fall through to env */
  }
  const url = (process.env.UNIFIED_BROKER_URL || '').replace(/\/+$/, '')
  const token = process.env.UNIFIED_BROKER_TOKEN || ''
  return url && token ? { url, token } : null
}

async function mintToken() {
  const coords = brokerCoords()
  if (!coords) return { token: null, expiresIn: 0 }
  try {
    const res = await fetch(`${coords.url}/token`, {
      method: 'POST',
      headers: { 'content-type': 'application/json', 'x-broker-token': coords.token },
      body: JSON.stringify({ app_slug: SLUG }),
    })
    if (!res.ok) {
      log(`broker token mint failed: HTTP ${res.status}`)
      return { token: null, expiresIn: 0 }
    }
    const data = await res.json()
    return data.token ? { token: data.token, expiresIn: Number(data.expires_in) || 300 } : { token: null, expiresIn: 0 }
  } catch (e) {
    log(`broker token mint error: ${e}`)
    return { token: null, expiresIn: 0 }
  }
}

function writeHostToken(token) {
  try {
    fs.writeFileSync(path.join(DATA_DIR, '.token'), token)
  } catch {
    /* best-effort handoff file */
  }
}

async function refreshLoop(expiresIn) {
  for (;;) {
    await new Promise((r) => setTimeout(r, Math.max(30, expiresIn - 60) * 1000))
    const { token, expiresIn: ttl } = await mintToken()
    if (token) {
      currentToken = token
      writeHostToken(token)
      expiresIn = ttl
      log('refreshed Unified app token')
    } else {
      expiresIn = 120 // back off and retry sooner on failure
    }
  }
}

// Loopback proxy: forwards an OpenAI-compatible request to the Unified gateway
// with the freshly-minted app token attached. Point an Activepieces AI provider's
// Base URL at http://127.0.0.1:<AI_PROXY_PORT>/v1 to use the Unified subscription.
function startAiProxy() {
  const server = http.createServer((req, res) => {
    if (!currentToken) {
      res.writeHead(401, { 'content-type': 'application/json' })
      res.end(JSON.stringify({ error: { message: 'Sign in to UnifiedApp to use AI on your Unified subscription.' } }))
      return
    }
    const subPath = (req.url || '/').replace(/^\/v1/, '') // accept base URL with or without /v1
    const target = new URL(UNIFIED_API_BASE + subPath)
    const transport = target.protocol === 'https:' ? https : http
    const headers = { ...req.headers, host: target.host, authorization: `Bearer ${currentToken}` }
    const upstream = transport.request(
      target,
      { method: req.method, headers },
      (up) => {
        res.writeHead(up.statusCode || 502, up.headers)
        up.pipe(res)
      },
    )
    upstream.on('error', (e) => {
      if (!res.headersSent) res.writeHead(502, { 'content-type': 'application/json' })
      res.end(JSON.stringify({ error: { message: `Unified gateway error: ${e}` } }))
    })
    req.pipe(upstream)
  })
  server.on('error', (e) => {
    log(`AI gateway proxy could not bind 127.0.0.1:${AI_PROXY_PORT} (${e.code}); AI-on-subscription disabled`)
  })
  server.listen(AI_PROXY_PORT, '127.0.0.1', () => {
    const baseUrl = `http://127.0.0.1:${AI_PROXY_PORT}/v1`
    try {
      fs.writeFileSync(path.join(DATA_DIR, '.unified-ai-gateway.json'), JSON.stringify({ baseUrl }))
    } catch {
      /* discovery file is best-effort */
    }
    log(`AI gateway proxy → ${UNIFIED_API_BASE}; set an Activepieces AI provider Base URL to ${baseUrl}`)
  })
}

// ── Children ──────────────────────────────────────────────────────────────────
const children = []
let shuttingDown = false

function spawnChild(name, entry, extraEnv) {
  const child = spawn(NODE_BIN, ['--enable-source-maps', entry], {
    cwd: BUNDLE_ROOT, // server.ts resolves dist/packages/web from process.cwd()
    env: { ...sharedEnv, ...extraEnv },
    stdio: 'inherit',
  })
  children.push(child)
  child.on('exit', (code, signal) => {
    log(`${name} exited (code=${code}, signal=${signal})`)
    // The app is the service the host health-checks; if it dies, surface failure.
    if (!shuttingDown && name === 'app') shutdown(code ?? 1)
  })
  child.on('error', (e) => log(`${name} failed to start: ${e}`))
  return child
}

function shutdown(code = 0) {
  if (shuttingDown) return
  shuttingDown = true
  for (const c of children) {
    try {
      c.kill('SIGTERM')
    } catch {
      /* already gone */
    }
  }
  setTimeout(() => process.exit(code), 3000).unref()
}
process.on('SIGTERM', () => shutdown(0))
process.on('SIGINT', () => shutdown(0))

async function main() {
  log(`Activepieces ${process.env.npm_package_version || ''} starting on ${BIND_HOST}:${PORT} (data=${DATA_DIR})`)

  const { token, expiresIn } = await mintToken()
  if (token) {
    currentToken = token
    writeHostToken(token)
    log(`acquired Unified app token (expires_in=${expiresIn}s) via broker`)
    refreshLoop(expiresIn).catch((e) => log(`token refresh loop stopped: ${e}`))
  } else {
    log('no broker token (running outside UnifiedApp?); AI gateway proxy will 401 until signed in')
  }
  startAiProxy()

  // app first (binds the port + runs migrations), then the worker (polls the app;
  // its socket client reconnects until the app is up, so ordering is not strict).
  spawnChild('app', APP_ENTRY, { AP_CONTAINER_TYPE: 'APP' })
  spawnChild('worker', WORKER_ENTRY, { AP_CONTAINER_TYPE: 'WORKER_AND_APP' })
}

main().catch((e) => {
  log(`fatal: ${e}`)
  process.exit(1)
})
